source/credentials_provider_sts_web_identity.c

```diff
@@ -187,128 +191,138 @@ Error Response looks like:
+static int s_stswebid_200_xml_on_root(struct aws_xml_node *node, void *user_data) {
+    struct aws_byte_cursor node_name = aws_xml_node_get_name(node);
+    if (aws_byte_cursor_eq_c_str_ignore_case(&node_name, "AssumeRoleWithWebIdentityResponse")) {
+        return aws_xml_node_traverse(node, s_stswebid_200_xml_on_AssumeRoleWithWebIdentityResponse_child, user_data);
+    }
+    return AWS_OP_SUCCESS;
+}
 
-    struct aws_byte_cursor node_name;
-    AWS_ZERO_STRUCT(node_name);
+static int s_stswebid_200_xml_on_AssumeRoleWithWebIdentityResponse_child(
+    struct aws_xml_node *node,
 
-    if (aws_xml_node_get_name(node, &node_name)) {
-        AWS_LOGF_ERROR(
-            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
-            "(id=%p): While parsing credentials xml response for sts web identity credentials provider, could not get "
-            "xml node name for function s_on_creds_node_encountered_fn.",
-            user_data);
-        return false;
+    void *user_data) {
+
+    struct aws_byte_cursor node_name = aws_xml_node_get_name(node);
+    if (aws_byte_cursor_eq_c_str_ignore_case(&node_name, "AssumeRoleWithWebIdentityResult")) {
+        return aws_xml_node_traverse(node, s_stswebid_200_xml_on_AssumeRoleWithWebIdentityResult_child, user_data);
     }
+    return AWS_OP_SUCCESS;
+}
+
+static int s_stswebid_200_xml_on_AssumeRoleWithWebIdentityResult_child(
+    struct aws_xml_node *node,
+
+    void *user_data) {
 
-    if (aws_byte_cursor_eq_c_str_ignore_case(&node_name, "AssumeRoleWithWebIdentityResponse") ||
-        aws_byte_cursor_eq_c_str_ignore_case(&node_name, "AssumeRoleWithWebIdentityResult") ||
-        aws_byte_cursor_eq_c_str_ignore_case(&node_name, "Credentials")) {
-        return aws_xml_node_traverse(parser, node, s_on_creds_node_encountered_fn, user_data);
+    struct aws_byte_cursor node_name = aws_xml_node_get_name(node);
+    if (aws_byte_cursor_eq_c_str_ignore_case(&node_name, "Credentials")) {
+        return aws_xml_node_traverse(node, s_stswebid_200_xml_on_Credentials_child, user_data);
     }
+    return AWS_OP_SUCCESS;
+}
 
+static int s_stswebid_200_xml_on_Credentials_child(struct aws_xml_node *node, void *user_data) {
     struct sts_web_identity_user_data *query_user_data = user_data;
+
+    struct aws_byte_cursor node_name = aws_xml_node_get_name(node);
     struct aws_byte_cursor credential_data;
     AWS_ZERO_STRUCT(credential_data);
+
     if (aws_byte_cursor_eq_c_str_ignore_case(&node_name, "AccessKeyId")) {
-        aws_xml_node_as_body(parser, node, &credential_data);
-        query_user_data->access_key_id =
-            aws_string_new_from_array(query_user_data->allocator, credential_data.ptr, credential_data.len);
+        if (aws_xml_node_as_body(node, &credential_data)) {
+            return AWS_OP_ERR;
+        }
+        query_user_data->access_key_id = aws_string_new_from_cursor(query_user_data->allocator, &credential_data);
```

Comment:
trivial: should we check the `access_key_id` is not empty? If it's empty, it will go all the way to when we create credentials to find out the error and report error while creating the credentials. So, maybe better to check it here and fail the parsing.

Reply:
I made it so we report `AWS_AUTH_CREDENTIALS_PROVIDER_STS_WEB_IDENTITY_SOURCE_FAILURE` if anything goes wrong parsing the response, and/or creating the credentials. That seems more useful than the AWS_ERROR_INVALID_ARGUMENT we would get with the old path, if the strings were ever empty.

Reply:
I decided to check at the end, that the credentials were successfully created, vs checking each element's string is non-empty. That way, if it ever becomes valid for one of these strings to be empty, we only have to fixed `aws_credentials_new()`, vs needing to fix everywhere that reads an XML element that eventually feeds into `aws_credentials_new()`